go/merge/three_way.go

```diff
@@ -120,95 +120,77 @@ func threeWayMapMerge(a, b, parent types.Map, vwr types.ValueReadWriter) (merged
 		b.DiffLeftRight(parent, bChangeChan, bStopChan)
 		close(bChangeChan)
 	}()
-	stopAndDrain := func(stop chan<- struct{}, drain <-chan types.ValueChanged) {
-		close(stop)
-		for range drain {
-		}
-	}
+
 	defer stopAndDrain(aStopChan, aChangeChan)
 	defer stopAndDrain(bStopChan, bChangeChan)
 
-	merged = parent
-	aChange, bChange := types.ValueChanged{}, types.ValueChanged{}
-	for {
-		// Get the next change from both a and b. If either diff(a, parent) or diff(b, parent) is complete, aChange or bChange will get an empty types.ValueChanged containing a nil Value. Generally, though, this allows us to proceed through both diffs in (key) order, considering the "current" change from both diffs at the same time.
-		if aChange.V == nil {
-			aChange = <-aChangeChan
-		}
-		if bChange.V == nil {
-			bChange = <-bChangeChan
+	apply := func(target types.Value, change types.ValueChanged, newVal types.Value) types.Value {
+		switch change.ChangeType {
+		case types.DiffChangeAdded, types.DiffChangeModified:
+			return target.(types.Map).Set(change.V, newVal)
+		case types.DiffChangeRemoved:
+			return target.(types.Map).Remove(change.V)
+		default:
+			panic("Not Reached")
 		}
+	}
+	return threeWayOrderedSequenceMerge(parent, aChangeChan, bChangeChan, a.Get, b.Get, parent.Get, apply, vwr)
+}
 
-		// Both channels are producing zero values, so we're done.
-		if aChange.V == nil && bChange.V == nil {
-			break
-		}
+func threeWayStructMerge(a, b, parent types.Struct, vwr types.ValueReadWriter) (merged types.Value, err error) {
+	aChangeChan, bChangeChan := make(chan types.ValueChanged), make(chan types.ValueChanged)
+	aStopChan, bStopChan := make(chan struct{}, 1), make(chan struct{}, 1)
 
-		// Since diff generates changes in key-order, and we never skip over a change without processing it, we can simply compare the keys at which aChange and bChange occurred to determine if either is safe to apply to the merge result without further processing. This is because if, e.g. aChange.V.Less(bChange.V), we know that the diff of b will never generate a change at that key. If it was going to, it would have done so on an earlier iteration of this loop and been processed at that time.
-		// It's also obviously OK to apply a change if only one diff is generating any changes, e.g. aChange.V is non-nil and bChange.V is nil.
-		if aChange.V != nil && (bChange.V == nil || aChange.V.Less(bChange.V)) {
-			merged = apply(merged, aChange, a.Get(aChange.V))
-			aChange = types.ValueChanged{}
-			continue
-		} else if bChange.V != nil && (aChange.V == nil || bChange.V.Less(aChange.V)) {
-			merged = apply(merged, bChange, b.Get(bChange.V))
-			bChange = types.ValueChanged{}
-			continue
-		}
-		d.PanicIfTrue(!aChange.V.Equals(bChange.V), "Diffs have skewed!") // Sanity check.
+	go func() {
```

Comment:
Can you move starting/stopping/draining the merge into the threeWayOrderedEtcMerge function? Like Diff. It would make it self contained.

The logic isn't trivial and it'd be good to have only a single instance of it.